### DISCUSSION TITLE
Use pdf tags for text extraction

### DISCUSSION THREAD

**Discussion by MartinThoma** [py-pdf/pypdf#1633] [Ideas]
The comments in https://stackoverflow.com/a/75419068/562769 might be interesting @pubpub-zz

**Comment by pubpub-zz** [comment]
It is interesting about the `/Span<</ActualText() >>`
@naourass can you try to give an explanation about the cleaning:
![image](https://user-images.githubusercontent.com/4083478/219495679-bb7b9af0-4074-4545-9e7c-4aa9eb54f82a.png)
When I decode the sequence I see no difference😯
```
s=b"\x06\x27\xFE\xDF\xFE\xE4\xFE\xEE\x06\x27\x06\x31\x06\x2F"
s.decode("utf-16-be")
'اﻟﻤﻮارد'
ss=b"\x06\x27\x06\x44\x06\x45\x06\x48\x06\x27\x06\x31\x06\x2F"
ss.decode("utf-16-be")
'الموارد'
````

**Reply by naourass** [reply]
@pubpub-zz 
In Arabic, there's [similar looking characters](https://en.wikipedia.org/wiki/Arabic_script_in_Unicode) that has multiple unicodes. Look the letter Ba or Ra for example:

![image](https://user-images.githubusercontent.com/4897498/219501243-418c4037-1fef-4475-bb89-2a87d5e5296d.png)
![image](https://user-images.githubusercontent.com/4897498/219501643-15786fa2-f79a-49e3-bc8a-7e46c4c2505d.png)


Often when these variants of the char are ligatured, they give the same visual result. And from what I experienced, sometimes just copying and pasting Arabic text in some editors may change the variant of some chars (maybe due to rich to plain text, normalization or some processing by the editor I'm not sure about the reason of the change).

**Reply by naourass** [reply]
BTW regarding the order, what I usually see in other text extraction/processing tools, is that they give you the logical order of the text and you need to call `bidi.algorithm.get_display()` to get the visual order.

https://unicode-org.github.io/icu/userguide/transforms/bidi.html
https://www.w3.org/International/articles/inline-bidi-markup/uba-basics

**Reply by naourass** [reply]
And for the cleaning, there's high chance that it could be just [normalization](https://docs.python.org/3/library/unicodedata.html#unicodedata.normalize).
